packages/contracts/contracts/authority/AuthorityFactory.sol

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {Authority} from "@cartesi/rollups/contracts/consensus/authority/Authority.sol";
+import {History} from "@cartesi/rollups/contracts/history/History.sol";
+import {Create2} from "@openzeppelin/contracts/utils/Create2.sol";
+
+import {IAuthorityFactory} from "./IAuthorityFactory.sol";
+
+/// @title Authority-History Pair Factory
+/// @notice Allows anyone to reliably deploy a new Authority-History pair.
+contract AuthorityFactory is IAuthorityFactory {
+    function newAuthority(
+        address _authorityOwner
+    ) external override returns (Authority) {
+        Authority authority = new Authority(address(this));
+        History history = new History(address(authority));
+        authority.setHistory(history);
+        authority.transferOwnership(_authorityOwner);
+
+        emit AuthorityCreated(authority);
+
+        return authority;
+    }
+
+    function newAuthority() external override returns (Authority) {
+        return this.newAuthority(msg.sender);
```

Comment:
Here, it is better to make the `newAuthority` you are calling a public function, and remove the `this.`. What this does is compile to a `JUMP` instead of a `CALL` opcode.

Reply:
By public, I mean it has the `public` modifier in its signature.
All `newAuthority` functions in this contract are external.
External functions can only be called externally (`CALL`).
Public functions can be called externally (`CALL`) and internally (`JUMP`).